presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -281,6 +291,49 @@ public ConnectorSplitSource getSplits(
                 layout.getPredicateColumns(),
                 layout.getDomainPredicate().getDomains());
 
+        double ratio = 1.0;
+
+        if (isDynamicSplitSizesEnabled(session)) {
+            HiveTableHandle hiveTableHandle = new HiveTableHandle(tableName.getSchemaName(), tableName.getTableName());
+            List<HiveColumnHandle> allColumnHandles = new ArrayList<>();
+            allColumnHandles.addAll(getRegularColumnHandles(table));
+            allColumnHandles.addAll(getPartitionKeyColumnHandles(table));
+            Map<String, Type> allColumnTypes = allColumnHandles.stream().collect(toImmutableMap(HiveColumnHandle::getName, handle -> metadata.getColumnMetadata(session, hiveTableHandle, handle).getType()));
+            Map<String, ColumnHandle> allColumns = allColumnHandles.stream().collect(toImmutableMap(HiveColumnHandle::getName, identity()));
+
+            Set<HiveColumnHandle> readColumnHandles = new HashSet<>();
+            if (layout.getRequestedColumns().isPresent()) {
+                readColumnHandles.addAll(layout.getRequestedColumns().get());
+            }
+            readColumnHandles.addAll(layout.getPredicateColumns().values());
+            readColumnHandles.removeIf(handle -> handle.isHidden());
+
+            TableStatistics tableStatistics = metadata.getHiveStatisticsProvider().getTableStatistics(session, tableName, allColumns, allColumnTypes, partitions);
+            double totalSize = tableStatistics.getTotalSize().getValue();
+            double requiredSize = 0;
+            double rowCount = tableStatistics.getRowCount().getValue();
+            for (Map.Entry<ColumnHandle, ColumnStatistics> entry : tableStatistics.getColumnStatistics().entrySet()) {
+                if (readColumnHandles.contains(entry.getKey())) {
+                    double value = entry.getValue().getDataSize().getValue();
+                    // We do not compute total size stats for fixed width types, so count them manually.
+                    if (!isFinite(value) && isFinite(rowCount)) {
+                        HiveColumnHandle columnHandle = (HiveColumnHandle) entry.getKey();
+                        Type type = metadata.getColumnMetadata(session, hiveTableHandle, columnHandle).getType();
+                        if (type instanceof FixedWidthType) {
+                            int size = ((FixedWidthType) type).getFixedSize();
+                            value = size * rowCount;
+                        }
+                    }
+                    requiredSize += value;
+                    readColumnHandles.remove(entry.getKey());
+                }
+            }
+
+            if (readColumnHandles.isEmpty() && totalSize > 0 && isFinite(totalSize) && isFinite(requiredSize)) {
```

Comment:
How could it happen that readColumnHandles.isEmpty() is false here?

Can we check totalSize before going over the loop?